### ISSUE TITLE
C++ static template methods: duplication as non-static and specialization

### ISSUE THREAD

**Issue by LRad09** [doxygen/doxygen#6986]
Using Doxygen 1.8.15, I ran into a couple of issues regarding templates and specialization.

Consider the following piece of code:
Foo.h:
```
class Foo
{
public:
   Foo();

private:
   /*!
      @brief Test method.

      @returns boolean value.
   */
   template<typename T>
   static bool Bar(int test /*!< [in] test paramater */);
};
```

Foo.cpp:
```
#include "Foo.h"

template<typename T>
bool Foo::Bar(T test)
{
   return (test == 0);
}

template<>
bool Foo::Bar(int test)
{
   return (test == 1);
}

template<>
bool Foo::Bar(std::string test)
{
   return (test == "");
}
```
Doxygen is making the following mistakes:
1. The template method is duplicated three times as public **and** not static (once per specialization). It should either be not duplicated or duplicated but as private and static, like the declaration of the method tells.
2. The documentation is **not** copied over to the specializations.
![private_static_template_methods](https://user-images.githubusercontent.com/50481138/57683553-3d47bf80-7634-11e9-8af7-7603e2283db6.png)

Declaring and defining the template method and its specializations in the .h file fixes the first issue (i.e. duplicated but as static private methods) but the documentation is still not copied over (this case is highlighted by the `FooInline` class in the attached archive).

Ideally, it would be nice to carry over the documentation from the base unspecialized method and to let the developer declare and define the specializations in the .cpp file.

I attached a zip archive with all the necessary files to reproduce the issue:
[static_template_specialization.zip](https://github.com/doxygen/doxygen/files/3181313/static_template_specialization.zip)

**Comment by albert-github** [comment]
Looks to me that you added the wrong zip file (looks to me the one from your previous issue).
Please correct.

**Comment by LRad09** [comment]
Indeed, I just edited my post with the following archive:
[static_template_specialization.zip](https://github.com/doxygen/doxygen/files/3181313/static_template_specialization.zip)

Regarding your question, this link [https://en.cppreference.com/w/cpp/language/template_specialization](https://en.cppreference.com/w/cpp/language/template_specialization) can give you more insight on template specialization in C++.

The first example given concerns `struct`s but as the list implies, member function template of a class can be specialized with that syntax.

**Comment by doxygen** [comment]
I've pushing [this commit](https://github.com/doxygen/doxygen/commit/9dad94dc6d3cb528cc7650cf07418a83ed551bdd) which should solve this issue.

**Comment by albert-github** [comment]
A fix has been integrated in master on github (please don't close the issue as this will be done at the moment of a release).

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.8.16.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
